### ISSUE TITLE
add maxLength and destroy 

### ISSUE THREAD

**Issue by dcolens** [myndzi/binary-split-streams2#1]
thanks for taking the time to write and publish this module, it does its job very well.

split/split2 have:
- the ability to pass a maxLength that will stop processing if a line is longer than maxLength
- a destroy method to stop the processing (is there a way to do that here?)

I have a use case where binary is required and I found this module does the best job at it, so I'd be interested to see the above 2 features added, is there any chance for adding those ?

**Comment by myndzi** [comment]
This module is just a transform stream, so you can unpipe or terminate the source stream and it will process to completion. If you want to stop processing from the split stream immediately, you can unpipe it from your destination, stop calling .read, or unbind any events. If you call .end on it, it will forbid further writes. Node 8 also introduces .destroy on writable streams (which should include transform streams).

Once the transform stream is not referenced it will be garbage collected like anything else.

There isn't any functionality to keep some kind of length counter, but that should be possible to add without too much trouble I think -- at least, if we're talking length in bytes. I think emitting an event makes the most sense here, maybe an 'error' event, though I'm rusty on the semantics of what that does with piped streams.

**Comment by myndzi** [comment]
@dcolens I've pushed code to allow for a configurable internal buffer size, and to emit a message when data is truncated. You should be able to test straight from the repo by using:

`npm install myndzi/binary-split-streams2`

I think this interface provides the tool you need with a generic and flexible interface that will allow you to both limit memory usage and do whatever you want in the case that you receive too much data without being split. This version also significantly increases the speed (>= twice as fast with a single byte splitter). Please let me know if this meets your needs. I'd like to ensure that is the case before publishing.

**Comment by myndzi** [comment]
I published it with the beta tag too, so you can also just `npm install binary-split-streams2@beta`

**Comment by dcolens** [comment]
Thanks, I'l test today!

**Comment by myndzi** [comment]
Ah, right, check the updated readme for example/usage: https://github.com/myndzi/binary-split-streams2/blob/master/README.md#truncation

**Comment by dcolens** [comment]
looks like truncated is emitted after the cumulated length > maxBuffer and not after one line exceeds maxBuffer ? If I set maxBuffer to anything less than the file size in my tests, it emits truncated (test file size is 4036 bytes, tested with values from 700 to 4035, max line length is < 700)

**Comment by myndzi** [comment]
maxBuffer controls the size of the buffer it stores internally to search over; the caveat here is this is done at the step where the buffer is accumulated, so it is possible that with a maxBuffer of 700 you will not receive a full line that is exactly 700 bytes long. If you want to always receive 700-byte lines, the best approach is to set maxBuffer to twice that (1400). The code would get pretty messy otherwise, and this seemed acceptable. I'm not sure I fully follow what you're saying, but does this answer your question?

In other words, the steps when receiving a new chunk have gone from:
- Add new chunk to internal buffer
- Search for splitters
to:
- Truncate internal buffer (or incoming chunk) if necessary
- Add new chunk to internal buffer
- Search for splitters

**Comment by myndzi** [comment]
(thinking about it, this may not be necessary - as long as the code is running, the new chunk and old buffer will be allocated, so it's possible I can truncate after searching; this would mean that lines longer than the specified value may be emitted, which could be ~fine?)

**Comment by dcolens** [comment]
Regarding 

> If you want to always receive 700-byte lines, the best approach is to set maxBuffer to twice that (1400).

Setting maxBuffers to twice the value also fails, actually setting it to anything anywhere between 0 and fileSize-1 fails (that's what I meant when writing I tried all sizes between 700 and 4035).

The only viable option at this stage is to check the size of the returned match, but it means that if the match is huge it might still deplete resources :/

**Comment by myndzi** [comment]
That doesn't seem correct, I'll look into it. Can you be more specific about what/how you're testing?

What should happen is that, after every chunk it reads, it should have <= maxBuffer bytes in internal storage. When it hits a splitter or the stream ends, it should emit or flush the remaining data. This does raise another interesting thought, which is that there's the potential in the current implementation to truncate a splitter, which is bad -- so I think I should probably modify it to truncate after searching instead of before.

**Comment by myndzi** [comment]
I just re-read what you wrote a few times, and I think what I mentioned above is the problem.

It truncates once per *chunk read*; if your file is small enough to be read in a single chunk, then it's cutting off everything one time regardless of the size. Fix incoming.

**Comment by myndzi** [comment]
Try `1.1.1@beta`. I also added an option, `strictTruncation: true`, which will ensure that all chunks you receive are <= maxBuffer in size if you want.

**Comment by myndzi** [comment]
@dcolens have you had a chance to check out the updated version?

**Comment by dcolens** [comment]
not yet, will try begin of next week!

**Comment by myndzi** [comment]
@dcolens did you get a chance to see if the changes work for you?

**Comment by myndzi** [comment]
Gonna just publish as-is; let me know if you have further problems
